yarn-project/foundation/src/retry/index.ts

```diff
@@ -38,6 +38,12 @@ export async function retry<Result>(
     try {
       return await fn();
     } catch (err: any) {
+      if (err.cause == 'thrownByServer') {
+        // If the error is specifically marked as thrown by the server, we don't retry because the error should be
+        // propagated. This is because it's an "intentional error" (e.g. "Contract is not deployed") and not
+        // a connection error or an unexpected software bug.
+        throw err;
+      }
```

Comment:
I don't think this belongs here: `retry` is a generic util, used in many different contexts, not just json-rpc client/server. It shouldn't know about this particular `cause`. I think the clean version to do it is either:
1. Include another optional param that takes an error and returns a bool on whether to abort retries or not
2. Have the function `fn` receive a parameter that can be used to bail, similar to how [async-retry](https://www.npmjs.com/package/async-retry) does it
3. Have `fn` wrap its non-retriable exceptions in a specific `NoRetryException`
4. Or have `fn` attach a `retry: false` property to the exception to avoid retries

Reply:
You have a good point and thank you for laying out the options so thoroughly. I think that given our specific use case, which is "forward all the errors to the client if the error was actually thrown by the server code", option 1 is too impractical (I want all the server thrown errors to be bubbled up so listing them all is too verbose) and option 2 is in my opinion unnecessarily generic (I don't need to do anything special with different errors, I just want it to not retry on server-thrown errors).

For this reason I think option 3 or 4 are the way to go.

I think option 3 will result in a more readable code so I will go with it.

Thank you for the feedback